### PR TITLE
Allow GetFrame to suspend frame rendering. 

### PR DESCRIPTION
**Pull request by stuxcrystal**
Implements #324 

Some rules: Do not suspend and request frames at the same time. On error, use cancelSuspension to clean up.

Also: Visual Studio can suck dicks, so please make sure that I didn't overlook any tabs Visual Studio might have added.

### CODE REVIEW

File: src/core/vsapi.cpp [hunk 1]
```diff
@@ -553,6 +553,26 @@ static void VS_CC getCoreInfo2(VSCore *core, VSCoreInfo *info) VS_NOEXCEPT {
     core->getCoreInfo2(*info);
 }
 
+static VSSuspensionContext* VS_CC suspendFrame(VSCore* core, VSFrameContext* frameCtx) VS_NOEXCEPT {
+    assert(core, frameCtx);
+	if (frameCtx->suspend) return 0;
```
Comment (FichteFoll) [inline src/core/vsapi.cpp]: tab

File: src/core/vsapi.cpp [hunk 2]
```diff
@@ -553,6 +553,26 @@ static void VS_CC getCoreInfo2(VSCore *core, VSCoreInfo *info) VS_NOEXCEPT {
...
+    assert(context);
+    context->core->threadPool->start(context->ctx);
+    delete context;
+}
+
+static void VS_CC cancelSuspend(VSSuspensionContext* context, VSFrameContext* frameCtx) VS_NOEXCEPT {
+	assert(context && frameCtx);
+	assert(context->ctx == frameCtx->ctx);
+	frameCtx->suspend = false;
+	delete context;
```
Comment (FichteFoll) [inline src/core/vsapi.cpp]: tabs

File: src/core/vsthreadpool.cpp [hunk 3]
```diff
@@ -216,10 +219,19 @@ void VSThreadPool::runTasks(VSThreadPool *owner, std::atomic<bool> &stop) {
 #ifdef VS_FRAME_REQ_DEBUG
             vsWarning("Exiting: %s Frame: %d Index: %d AR: %d Req: %d", mainContext->clip->name.c_str(), mainContext->n, mainContext->index, (int)ar, (int)mainContext->reqOrder);
 #endif
+			bool suspending = externalFrameCtx.suspend;
```
Comment (FichteFoll) [inline src/core/vsthreadpool.cpp]: tabs

**Comment by HolyWu** [comment]
I would recommend users of Visual Studio simply changing the setting in **Tools -> Option -> Text Editor -> C/C++ -> Tabs** from **Keep tabs** to **Insert spaces** once and for all.

**Comment by myrsloik** [comment]
I'll get to this one but it'll probably take a while.

**Comment by sekrit-twc** [comment]
How does this feature work? Can you show an example sequence of API calls that a "suspended" filter would make?

**Comment by stuxcrystal** [comment]
Of course.

Here is a completely useless example:
https://gist.github.com/stuxcrystal/a9867f17a44663330c1ae686c4a08c8d

**Comment by sekrit-twc** [comment]
Why do we need a "VSSuspensionContext" type? Can't the frame context track if it 's suspended?

**Comment by stuxcrystal** [comment]
Yes.

VSSuspensionContext is not used anywhere inside vapoursynth itself. It acts more like a token for the API-Consumer and is freed once the token is used (either by being cancelled or by resuming the frame). This way we can be sure that frame is only resumed once and cannot be used in subsequent suspensions of the same frame.

**Comment by sekrit-twc** [comment]
Is there a limit to the number of suspended operations? It looks like filters may have to be aware of this to avoid overwhelming the underlying resource.

**Comment by stuxcrystal** [comment]
There is no limit of suspended operations. Filters are free to use queues and semaphores for avoiding the underlying resource.

Renderers like vspipe and `VideoNode.output` only request a limited number of frames concurrently as they otherwise would request all frames at once thereby completly filling ram (This scenario would already happen, just try to request all frames at once using successive `VideoNode.get_frame_async`-calls for all frames on python.)

**Comment by stuxcrystal** [comment]
I don't want to rush but do we get an ETA for your decision/input, myrsloik?

**Comment by myrsloik** [comment]
Basically I'm hoping to get a first draft of audio support in before this and see where that goes. Unfortunately people keep giving me money to actually do other things...

**Comment by myrsloik** [comment]
One more reason I remembered: It'd be nice to implement cancellation of frame requests so vapoursynth can abort properly if the underlying resource gets stuck. So that probably has to go in first as well.

**Comment by stuxcrystal** [comment]
I kinda see no direct connection between audio and this.

The latter one is understandable though. Are you working on cancelation right now?

**Comment by myrsloik** [comment]
No, audio first. Actually I spend every minute working on paid stuff now because of deadlines.

**Comment by myrsloik** [comment]
Too many changes to merge this now. Will lump it into some general improvement issues instead.

**Pull request closed without merging**